rustls/src/crypto/aws_lc_rs/hpke.rs

```diff
@@ -0,0 +1,774 @@
+        let eae_prk = labeled_extract(self.hkdf, &suite_id, None, b"eae_prk", dh);
+        //   "For the variants of DHKEM defined in this document, the size Nsecret of the KEM shared
+        //   secret is equal to the output length of the hash function underlying the KDF."
+        let nsecret = u16::try_from(eae_prk.hash_len()).map_err(|_| OutputLengthError)?;
+
+        labeled_expand(&suite_id, eae_prk, b"shared_secret", kem_context, nsecret)
+    }
+}
+
+/// KeySchedule holds the derived base mode HPKE key material, base nonce, and seq number
+/// common to both a [Sealer] and [Opener].
+struct KeySchedule {
+    aead: &'static aead::Algorithm,
+    key: Vec<u8>,
+    base_nonce: Vec<u8>,
+    seq_num: u32,
+}
+
+impl KeySchedule {
+    /// See RFC 9180 §5.2 "Encryption and Decryption"[^0].
+    ///
+    /// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-5.2
+    fn compute_nonce(&self) -> Vec<u8> {
+        // def Context<ROLE>.ComputeNonce(seq):
+        //   seq_bytes = I2OSP(seq, Nn)
+        //   return xor(self.base_nonce, seq_bytes)
+
+        // Each new N-byte nonce is conceptually two parts:
+        //   * N-4 bytes of the base nonce (0s in `nonce` to XOR in as-is).
+        //   * 4 bytes derived from the sequence number XOR the base nonce.
+        let mut nonce = vec![0_u8; self.base_nonce.len()];
+        let seq_bytes = self.seq_num.to_be_bytes();
+        nonce[self.base_nonce.len() - seq_bytes.len()..].copy_from_slice(&seq_bytes);
+
+        for (n, &b) in nonce.iter_mut().zip(&self.base_nonce) {
+            *n ^= b;
+        }
+
+        nonce
+    }
+
+    /// See RFC 9180 §5.2 "Encryption and Decryption"[^0].
+    ///
+    /// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-5.2
+    fn increment_seq_num(&mut self) -> Result<(), aws_lc_rs::error::Unspecified> {
+        // def Context<ROLE>.IncrementSeq():
+        //   if self.seq >= (1 << (8*Nn)) - 1:
+        //     raise MessageLimitReachedError
+        //   self.seq += 1
+
+        // Determine the maximum sequence number using the AEAD nonce's length in bits.
+        // Do this as an u128 to prevent overflowing.
+        let max_seq_num = (1u128 << (self.aead.nonce_len() * 8)) - 1;
+
+        // Promote the u32 sequence number to an u128 and compare against the maximum allowed
+        // sequence number.
+        if u128::from(self.seq_num) >= max_seq_num {
+            return Err(aws_lc_rs::error::Unspecified);
+        }
+
+        self.seq_num += 1;
+        Ok(())
+    }
+}
+
+impl NonceSequence for &mut KeySchedule {
+    fn advance(&mut self) -> Result<AwsNonce, aws_lc_rs::error::Unspecified> {
+        let nonce = self.compute_nonce();
+        self.increment_seq_num()?;
+        AwsNonce::try_assume_unique_for_key(&nonce)
+    }
+}
+
+/// See RFC 9180 §4 "Cryptographic Dependencies"[^0].
+///
+/// [0]: https://www.rfc-editor.org/rfc/rfc9180.html#section-4
+fn labeled_extract(
+    hkdf: &'static dyn Hkdf,
+    suite_id: &[u8],
```

Comment:
That's a lot of `&[u8]` arguments. Is there more type safety to be had? It looks like `label` is arguably a `&str` and `suite_id` is actually a `HpkeSuiteId`...

Reply:
I'm not sure about label being a string. Perhaps 
```rust
enum Label { Key, Secret, ... };

impl AsRef<[u8]> for Label {
  fn as_ref(&self) -> &[u8] {
    match self { Label::Key => &b"key"[..], ... }
  }
}
```?

Reply:
Sounds reasonable to me. I was matching the RFC too closely here I think and could introduce some more type safety on top.

Reply:
Implemented the `Label` enum.

Reply:
What about `suite_id`? It looks like that could/should be a `&HpkeSuite`.

Reply:
We'd need more than just the `&HpkeSuite` because the value used has a different prefix/encoding depending on context (quoting the bottom of [9180, §4](https://www.rfc-editor.org/rfc/rfc9180.html#section-4)):

> The value of suite_id depends on where the KDF is used; it is assumed implicit from the implementation and not passed as a parameter. If used inside a KEM algorithm, suite_id MUST start with "KEM" and identify this KEM algorithm; if used in the remainder of HPKE, it MUST start with "HPKE" and identify the entire ciphersuite in use. 

Right now the callers are constructing the correct form as passing `&[u8]`, but I think another enum like `Label` would be better. Something like:

```rust
enum LabeledSuiteId {
  Hpke(HpkeSuite),
  Kem(HpkeKem),
}
```

Reply:
Suite ID fixed up to use an enum.